MaskFormer Swin-B (TTNN) bring-up + PCC + perf (N300)

bro4all opened the pull request:
## Summary

MaskFormer Swin-B (`facebook/maskformer-swin-base-coco`) ported to full TTNN on **Wormhole N300** (single chip).

All four components run on-device: Swin backbone, pixel decoder, transformer decoder, class + mask heads.

## Performance (N300, 320x320, batch=1)

| | Latency | FPS | vs Stage 1 |
|---|---------|-----|------------|
| **Stage 1** (bring-up) | 1069.62 ms | 0.93 | — |
| **Stage 2** (basic opt) | 934.30 ms | 1.07 | +12.6% |
| **Stage 3** (deeper opt, no cache) | 904.63 ms | 1.11 | +15.4% |
| **Stage 3 + program cache** | **71.60 ms** | **13.97** | **93.3%** |
| Trace+2CQ replay | 64.49 ms | 15.50 | — |
| Pure device compute | 62.85 ms | 15.91 | — |

### Stage 3 optimizations

- **ROW_MAJOR layer norm**: pass ROW_MAJOR tensors directly to `ttnn.layer_norm`, eliminating ~152 TILE-ROW_MAJOR layout conversion ops per forward across all backbone layer norms.
- **Direct K transpose**: pre-transpose K during QKV permute to remove a redundant permute per attention block.
- **Fused 2D spatial roll**: combine two sequential 1D rolls in shifted-window attention into a single quadrant-slice + concat pass.
- **Zero-copy view reshapes**: use `ttnn.view` for contiguous MLP and patch-merging reshapes.
- **Program cache enabled**: stable on tuned N300 path; cuts repeated-forward recompilation.
- **Decoder SDPA enabled**: lowered min-seq threshold to 96 so the 100-query decoder takes the SDPA fast path.

### PCC (4/4 tests pass)

- Class logits: `0.999602`
- Mask logits: `0.992698`
- Threshold: `0.97`

### COCO accuracy (Stage 1, 200 images)

- mIoU: `0.469`
- PQ: `0.383`

## Repro

```bash
# Stage 3 demo (320x320)
python -m models.experimental.maskformer_swin.demo.runner \
  --image models/sample_data/demo.jpeg \
  --weights facebook/maskformer-swin-base-coco \
  --device wormhole_n300 \
  --height 320 --width 320 \
  --optimization-stage stage3 \
  --tt-repeats 5 \
  --dump-perf generated/maskformer_swin/stage3_opt/perf.json \
  --dump-perf-header generated/maskformer_swin/stage3_opt/perf_header.json \
  --output-dir generated/maskformer_swin/stage3_opt/demo_outputs

# PCC test
PYTHONPATH=$(pwd) pytest models/experimental/maskformer_swin/tests/pcc/test_maskformer_swin.py -q

# E2E trace+2CQ perf test
PYTHONPATH=$(pwd) pytest models/experimental/maskformer_swin/tests/perf/test_perf_e2e_maskformer_swin.py -q
```

<details>
<summary>All repro commands (stages 1-3, COCO eval, Tracy profiling)</summary>

### Stage 1 demo
```bash
python -m models.experimental.maskformer_swin.demo.runner \
  --image models/sample_data/demo.jpeg \
  --weights facebook/maskformer-swin-base-coco \
  --device wormhole_n300 \
  --height 320 --width 320 \
  --optimization-stage stage1 \
  --tt-repeats 5 \
  --dump-perf generated/maskformer_swin/stage1_baseline/perf.json \
  --dump-perf-header generated/maskformer_swin/stage1_baseline/perf_header.json \
  --output-dir generated/maskformer_swin/stage1_baseline/demo_outputs
```

### Stage 2 demo
```bash
python -m models.experimental.maskformer_swin.demo.runner \
  --image models/sample_data/demo.jpeg \
  --weights facebook/maskformer-swin-base-coco \
  --device wormhole_n300 \
  --height 320 --width 320 \
  --optimization-stage stage2 \
  --tt-repeats 5 \
  --dump-perf generated/maskformer_swin/stage2_opt/perf.json \
  --dump-perf-header generated/maskformer_swin/stage2_opt/perf_header.json \
  --output-dir generated/maskformer_swin/stage2_opt/demo_outputs
```

### COCO eval (50 images)
```bash
python -m models.experimental.maskformer_swin.demo.runner \
  --image models/sample_data/demo.jpeg \
  --weights facebook/maskformer-swin-base-coco \
  --device wormhole_n300 \
  --height 320 --width 320 \
  --optimization-stage stage1 \
  --tt-repeats 1 \
  --output-dir generated/maskformer_swin/stage1_baseline/demo_outputs \
  --coco-eval \
  --coco-dir generated/maskformer_swin/coco_subset \
  --coco-max-images 50 \
  --coco-report generated/maskformer_swin/stage1_baseline/coco_eval/report_50.json
```

### Tracy perf sheet
```bash
PYTHONPATH=$(pwd)/ttnn \
TT_METAL_LOGS_PATH=/tmp/tt_logs \
TT_METAL_INSPECTOR=0 \
TT_METAL_PROFILER_PROGRAM_SUPPORT_COUNT=20000 \
TT_METAL_PROFILER_MID_RUN_DUMP=1 \
./tools/tracy/profile_this.py \
  -o generated/maskformer_swin/stage3_opt/perf_sheet \
  -n maskformer_swin_stage3_320 \
  -c "models.experimental.maskformer_swin.demo.runner --image models/sample_data/demo.jpeg --weights facebook/maskformer-swin-base-coco --device wormhole_n300 --height 320 --width 320 --optimization-stage stage3 --tt-repeats 1 --output-dir generated/maskformer_swin/stage3_opt/profile_run"
```

</details>

## Artifacts

- Perf sheet gist: https://gist.github.com/bro4all/426d9f41889c9b80ac462c276cfcf96c
- Demo outputs: `generated/maskformer_swin/stage1_baseline/demo_outputs/` (semantic overlay, panoptic map, instance masks)

## Notes

- Single chip only (`device_id=0`); not validated on N150.
- Platform auto-detected and written to `perf_header.json`.
- Full-TT execution enforced (no TT-to-torch-to-TT detours).

marty1885 commented:
@bro4all As per requirement for submission in the issue

* Performance report + header for final review

Please produce the performance result and headers. See

https://github.com/tenstorrent/tt-metal/blob/main/tech_reports/ttnn/TTNN-model-bringup.md#41-performance-sheet
https://docs.tenstorrent.com/tt-metal/latest/ttnn/ttnn/profiling_ttnn_operations.html#perf-report-headers

We cannot start the review until the document is generated

bro4all commented:
[ops_perf_results_2025_11_18_20_45_58.csv](https://github.com/user-attachments/files/23615030/ops_perf_results_2025_11_18_20_45_58.csv)
@marty1885

bro4all commented:
@marty1885

bro4all committed:
MaskFormer Swin-B TTNN implementation

Port MaskFormer-Swin-B segmentation model to TTNN for Wormhole N300.

Components on Tensix:
- Transformer decoder with TTNN attention and fused QKV+SDPA
- Classification and mask projection heads with program-configured matmuls
- Optional 3×3 mask projection convolution (TT-CNN path with CPU fallback)

Components on CPU:
- Swin-B backbone (patch embedding + 4 stages)
- Pixel decoder (FPN-style upsampling)

Features:
- Hybrid TT/CPU execution with `--tt-run` flag
- CPU-only fallback with `--fallback-run` for validation
- Supports 320×320 and 640×640 input sizes
- COCO evaluation integration (mIoU and optional PQ)
- Comprehensive test suite with parity checks

Closes #30876

bro4all committed:
MaskFormer Swin-B: perf artifacts and sample outputs

Add performance measurements and example outputs for Wormhole N300:

Performance headers:
- perf_header_wormhole_n300.json: Native resolution (1365×2048)
- perf_header_wormhole_n300_320.json: 320×320 (~1 FPS)
- perf_header_wormhole_n300_640.json: 640×640 (~0.29 FPS)

Sample outputs (JSON only, overlay image excluded due to size):
- tt_perf_demo.json: Example performance metrics
- coco_eval_demo.json: Example COCO evaluation results

bro4all committed:
fix: remove l1_small_size override for device compatibility

bro4all committed:
maskformer_swin: update perf measurements after l1_small_size fix

Regenerated performance measurements on Wormhole N300:
- 320x320: 1077.6ms latency (0.93 FPS)
- 640x640: 3431.4ms latency (0.29 FPS)

bro4all committed:
maskformer_swin: move perf/eval artifacts to PR comment; delete repo copies and update README

bro4all committed:
maskformer_swin: restructure into demo/, tt/, tests/ subdirectories

Reorganize file structure to match the approved pattern for experimental
models (consistent with swin_s, vgg, dpt_large):

- Move implementation files to tt/ subdirectory
- Move runner.py to demo/ subdirectory
- Update all imports to reflect new paths
- Update README.md runner command paths

minaliuTT commented:
> @minaliuTT can I please get some traction on this to get paid eventually ?

@dgomezTT @jmalone-tt Could you please take a look when you have a moment? 
@jberkowitzTT Perf artifacts were uploaded above, could you check if there's anything else needed?

Thanks!

bro4all commented:
@jberkowitzTT Thanks for the heads up! I investigated the CI failure with Codex.

## Root Cause
The CI failure is **not caused by the PR code** - it's a CI infrastructure timing issue:

```
CMake Error at tt_metal/hw/CMakeLists.txt:127 (message):
  Installed SFPI version 7.15.0 does not match required version 7.16.0
```

**What happened:**
1. SFPI 7.16.0 was merged to main yesterday (commit `cd24e508a7`, Jan 14 20:42 UTC) via PR #35849
2. The CI Docker images still have SFPI 7.15.0 installed
3. When CI runs on any PR, it creates a merge commit with current main, which now requires SFPI 7.16.0
4. The version mismatch causes the wheel build to fail

## Verification
- This PR only touches `models/experimental/maskformer_swin/` (model code only)
- No build, CMake, or infrastructure files are modified
- The same failure would affect any PR targeting main until the CI Docker images are updated

## Actions Taken
- Rebased onto latest `upstream/main` (now at `9fa927f9c7`) to stay current
- Pushed updated branch

## Resolution
This requires the CI team to rebuild the wheel-build Docker images (`manylinux-amd64`) with SFPI 7.16.0. Once the Docker images are updated, CI should pass.

Is there anything else I can do on my end, or does this need to wait for the Docker image rebuild?

bro4all committed:
perf: parse device perf CSV without branch header

bro4all commented:
@jberkowitzTT thanks — I updated the branch.

- Rebased `feature/maskformer-swin-ttnn-bounty` onto the latest `main` (as of Jan 16, 2026) to address the “out of date” warning.
- Investigated the current failing CI check: `trigger-post-commit / ops-unit-tests (wormhole_b0, N300) / wormhole_b0 sdxl op tests`.
  - All the SDXL perf tests themselves pass, but the job fails afterwards with `StopIteration` in `models/perf/perf_utils.py::process_perf_results` when parsing `Ops_Perf.csv` on the PR merge ref (detached HEAD, no `branch:` line).
  - I pushed a small fix to make the perf CSV parser handle the detached-head header layout.

New head SHA: `51cd4dfc6382319f30e3daca7bd7adfa33892a3d`.

Could you (or another org member) please rerun the post-commit workflows on the updated head to confirm CI is green now?

jberkowitzTT commented:
yup, just re-ran!

jberkowitzTT commented:
@jmalone-tt - could you please review this PR when you have the chance?

Thanks!

bro4all committed:
maskformer_swin: trim to required demo + reporting

bro4all committed:
Revert perf_utils CSV parsing change

bro4all committed:
maskformer_swin: trim runner and remove review clutter

bro4all committed:
maskformer_swin: full TTNN model + PCC/perf tests

bro4all commented:
@jberkowitzTT any updates here

jberkowitzTT commented:
/codeowners ping

dvartaniansTT commented on models/experimental/maskformer_swin/tests/perf/test_perf.py:
this pytest failed for me testing it on N150. also no instructions in the README on how to execute the perf test.

dvartaniansTT reviewed (changes requested):
please add instructions in the README on how to run all tests and what to expect from the test output.

dvartaniansTT reviewed (changes requested):
pytests are failing and README needs enhancements.

bro4all committed:
maskformer_swin: stabilize TTNN runtime compatibility and artifact export

bro4all committed:
maskformer_swin: harden pixel-decoder groupnorm execution paths

bro4all committed:
maskformer_swin: runner artifacts, N300 cache workaround, coco subset eval

bro4all committed:
maskformer_swin: fix pixel decoder ReLU after GroupNorm

bro4all committed:
maskformer_swin: fix decoder q scaling and harden TT heads

bro4all committed:
maskformer_swin: backbone caching and compute kernel controls

bro4all committed:
maskformer_swin: fix COCO eval resizing and category id mapping

bro4all committed:
maskformer_swin: optimize backbone and decoder attention

bro4all committed:
maskformer_swin: update README and fix perf test

bro4all committed:
maskformer_swin: stage3 backbone attention optimizations

bro4all committed:
maskformer_swin: cleanup perf test + remove global monkeypatch

bro4all committed:
maskformer_swin: strip bidi unicode from README

bro4all committed:
maskformer_swin: align caching defaults + docs

bro4all commented:
Uploaded fresh Tracy perf-sheet artifacts from a new N300 Stage-3 run on **February 19, 2026**.

### Attached artifacts
- Ops perf sheet (CSV): https://gist.github.com/bro4all/426d9f41889c9b80ac462c276cfcf96c
  - File: `Ops_Perf.csv` (271 lines)
- Repro command used: included in gist as `command.txt`

### Command used
```bash
PYTHONPATH=$(pwd)/ttnn \
TT_METAL_PROFILER_PROGRAM_SUPPORT_COUNT=20000 \
TT_METAL_PROFILER_MID_RUN_DUMP=1 \
./tools/tracy/profile_this.py \
  -o generated/maskformer_swin/stage3_opt/perf_sheet \
  -n maskformer_swin_stage3_320 \
  -c "models.experimental.maskformer_swin.demo.runner --image models/sample_data/demo.jpeg --weights facebook/maskformer-swin-base-coco --device wormhole_n300 --height 320 --width 320 --optimization-stage stage3 --tt-repeats 1 --output-dir generated/maskformer_swin/stage3_opt/profile_run"
```

### Notes / caveat
- The environment still shows the intermittent local profiler/JIT issue around `chlkc_descriptors.h` during kernel JIT.
- First pass produced `generated/maskformer_swin/stage3_opt/perf_sheet/Ops_Perf.csv` and a timestamped device-report folder at `generated/maskformer_swin/stage3_opt/perf_sheet/reports/maskformer_swin_stage3_320/2026_02_19_04_23_42/`.
- A second pass retried immediately but reproduced the same instability and reported no device logs.

If needed, I can rerun once the local JIT/profiler environment is stabilized to attach a clean device-only ops report as well.

bro4all committed:
maskformer_swin: enforce full TT path and update repro docs

bro4all committed:
maskformer_swin: fix pre-commit black formatting

dvartaniansTT commented on models/experimental/maskformer_swin/README.md:
```diff
@@ -0,0 +1,220 @@
+  --weights facebook/maskformer-swin-base-coco \
+  --device wormhole_n300 \
+  --height 320 --width 320 \
+  --optimization-stage stage3 \
```
Testing on N150, I get `TT run latency: mean=2733.35 ms std=38.27 ms (0.37 FPS over 5 runs)` which is less than the reported perf of `1149.33 ms` for stage 3 and 5 runs!

dvartaniansTT reviewed (changes requested):
Great comprehensive work. 

- Please confirm if this is running on single chip on N300? 
- I have tested on N150 and I see much larger latency for stage 3 compared to the reported numbers in README. please see my comments on README
- The end to end pytest should include trace+2cq implementation for ideal end to end perf. see an example here: https://github.com/bro4all/tt-metal/tree/d8f5ff4afe4a0ebd67970b054f82adfb76cd71fb/models/demos/vision/classification/mobilenetv2#testing

bro4all committed:
maskformer_swin: add e2e trace+2cq perf test and trace fixes

bro4all commented:
@dvartaniansTT 
 Confirmed: this is running on a single Wormhole chip on N300 (not data-parallel). The runner explicitly opens device_id=0, and perf_header.json now includes the detected platform/mesh (cluster_type,
 mesh_num_devices, mesh_device_ids, available_device_ids, opened_device_id) to make this unambiguous.

Re the N150 latency you saw (2733 ms vs README 1149 ms): the README numbers are measured on N300 single chip, so they aren’t expected to match N150. Also, --device wormhole_n300 is a reporting label only (it
 doesn’t select hardware). The runner now detects the actual platform and will warn if the passed label doesn’t match, and it records both device_label_arg and detected cluster_type in perf_header.json so
 comparisons are apples-to-apples.

 I only have N300 access on Koyeb, so I can’t reproduce N150 locally. If you can paste your N150 perf_header.json (and the “Harvesting masks …” line / fw bundle version printed on device open), I can help
  investigate whether the delta is from harvesting / fw+driver / host setup vs something unexpected.

  Per your suggestion, I added an “ideal E2E perf” path using trace replay + 2 command queues: pytest models/experimental/maskformer_swin/tests/perf/test_perf_e2e_maskformer_swin.py -q (README updated).

bro4all committed:
maskformer_swin: black format trace+2cq files

bro4all committed:
maskformer_swin: stage3 deeper optimizations — ROW_MAJOR layer norm, direct K transpose, program cache

Optimize Stage 3 inference latency on N300 (320x320, single chip):

Backbone attention:
- Eliminate TILE↔ROW_MAJOR layout conversion round-trips around all
  layer norms by passing ROW_MAJOR tensors directly to ttnn.layer_norm
  (saves ~152 to_layout ops per forward across block loop, output norms,
  and patch merging).
- Pre-transpose K during QKV permute (0,2,3,1 instead of 0,2,1,3)
  to eliminate a redundant permute in the attention score path.
- Use ttnn.view for contiguous reshapes in MLP and patch merging
  (zero-copy buffer reinterpretation vs data-copying reshape).
- Fuse the two sequential 1D spatial rolls in shifted-window attention
  into a single 2D quadrant-slice + concat pass that reads the input
  tensor once instead of twice.

Runner / stage3 config:
- Enable program cache on the tuned stage3 N300 path
  (MASKFORMER_TT_DISABLE_PROGRAM_CACHE=0) — stable and cuts
  repeated-forward recompilation cost substantially.
- Enable decoder SDPA (MASKFORMER_TT_ENABLE_SDPA=1) with lowered
  minimum sequence threshold (96) so the 100-query decoder path
  takes the SDPA fast path.
- Preserve explicit MASKFORMER_TT_* env overrides instead of
  unconditionally overwriting them with stage defaults.

Measured results (N300, 320x320, batch=1, program cache ON):
- Demo (5-repeat avg): 71.60 ms  (13.97 FPS)
- Trace+2CQ replay:    64.49 ms
- Pure device compute:  62.85 ms
- PCC: class=0.999602, mask=0.992698 (4/4 tests pass)

SeanNijjar commented:
this has been open for a while. Can we help @bro4all ?

SeanNijjar commented:
/codeowners ping
